File: Assets/Game/Scripts/Actor/ActorTriggerEffect.cs
//------------------------------------------------------------------------------
// This file is part of MistLand project in Nirvana.
// Copyright © 2016-2016 Nirvana Technology Co., Ltd.
// All Right Reserved.
//------------------------------------------------------------------------------

using System;
using System.Collections;
using System.Collections.Generic;
using Nirvana;
using UnityEngine;

/// <summary>
/// Used to play effect when animation event triggered.  
/// </summary>
[Serializable]
public sealed class ActorTriggerEffect : ActorTriggerBase
{
    private static Nirvana.Logger logger =
        LogSystem.GetLogger("ActorTriggerEffect");

    [SerializeField]
    [Tooltip("The effect prefab")]
    [AssetType(typeof(EffectControl))]
    private AssetID effectAsset;

    [SerializeField]
    [Tooltip("Whether to play at target.")]
    private bool playAtTarget;

    [SerializeField]
    [Tooltip("The reference node for the effect to play.")]
    private Transform referenceNode;

    [SerializeField]
    [Tooltip("Whether attach to the reference node.")]
    private bool isAttach;

    [SerializeField]
    [Tooltip("Whether rotation to the reference node.")]
    private bool isRotation;

    [SerializeField]
    [ActorEvent]
    [Tooltip("The event name stop this effect.")]
    private string stopEvent;

    // The stop event handle.
    private SignalHandle stopHandle;

    // All running effect.
    private LinkedList<EffectControl> effects =
        new LinkedList<EffectControl>();

    public AssetID EffectAsset
    {
        get { return this.effectAsset;  }
        set { this.effectAsset = value;  }
    }

    public bool PlayerAtTarget { get { return playAtTarget; } }

    public Transform ReferenceNode { get { return referenceNode; } }

    public bool IsAttach { get { return isAttach; } }

    public bool IsRotation { get { return isRotation; } }

    public string StopEvent { get { return stopEvent; } }

    /// <summary>
    /// Stop all effects.
    /// </summary>
    public void StopEffects()
    {
        foreach (var effect in this.effects)
        {
            if (effect != null)
            {
                effect.Stop();
            }
        }
    }

    /// <inheritdoc/>
    public override void Init(
        ActorTriggers actorTriggers, 
        AnimatorEventDispatcher dispatcher, 
        IActorTarget target)
    {
        base.Init(actorTriggers, dispatcher, target);
        this.stopHandle = dispatcher.ListenEvent(
            this.stopEvent,
            (param, stateInfo) =>
            {
                this.StopEffects();
            });
    }

    /// <inheritdoc/>
    public override void Destroy()
    {
        base.Destroy();
        if (this.stopHandle != null)
        {
            this.stopHandle.Dispose();
            this.stopHandle = null;
        }
    }

    /// <inheritdoc/>
    protected override void OnEventTriggered(
        Transform source, Transform target, AnimatorStateInfo stateInfo)
    {
        Scheduler.RunCoroutine(this.OnEventTriggeredImpl(
            source, target, stateInfo));
    }

    private IEnumerator OnEventTriggeredImpl(
        Transform source, Transform target, AnimatorStateInfo stateInfo)
    {
        if (this.effectAsset.IsEmpty)
        {
            logger.LogWarning("Missing effect trigger's prefab.");
            yield break;
        }

        // Find the reference node.
        Transform reference = null;
        Transform deliverer = null;
        if (this.playAtTarget)
        {
            reference = target;
            deliverer = (this.referenceNode != null) ?
                this.referenceNode : source.transform;
        }
        else
        {
            reference = (this.referenceNode != null) ?
                this.referenceNode : source.transform;
            deliverer = (this.referenceNode != null) ?
              this.referenceNode : source.transform;
        }

        if (reference == null || deliverer == null)
        {
            yield break;
        }

        // Load the effect.
        var wait = GameObjectPool.Instance.SpawnAsset(this.effectAsset);
        yield return wait;

        if (this == null)
        {
            yield break;
        }

        if (wait.Error != null)
        {
            logger.LogWarning(
                "Load prefab {0} failed: {1}.", this.effectAsset, wait.Error);
            yield break;
        }

        // Instantiate this effect.
        var instance = wait.Instance;
        if (instance == null)
        {
            logger.LogWarning(
                "The prefab {0} get instance from pool failed.", this.effectAsset);
            yield break;
        }

        instance.SetLayerRecursively(this.Actor.layer);
        var effect = instance.GetComponent<EffectControl>();
        if (effect == null)
        {
            logger.LogWarning(
                "The prefab {0} does not has EffectControl.", this.effectAsset);
            yield break;
        }

        // Setup the transform.
        if (this.isAttach)
        {
            effect.transform.SetParent(reference);
            if (this.isRotation)
            {
                var direction = reference.position - deliverer.position;
                direction.y = 0.0f;
                effect.transform.SetPositionAndRotation(
                    reference.position, Quaternion.LookRotation(direction));
            }
            else
            {
                effect.transform.localPosition = Vector3.zero;
                effect.transform.localRotation = Quaternion.identity;
            }
            effect.transform.localScale = reference.localScale;
        }
        else
        {
            effect.transform.SetPositionAndRotation(
                reference.position, reference.rotation);
            effect.transform.localScale = reference.lossyScale;
        }

        // Record and check the effect finish.
        var node = this.effects.AddLast(effect);
        effect.FinishEvent += () =>
        {
            this.effects.Remove(node);
            GameObjectPool.Instance.Free(effect.gameObject);
        };

        // Start to play.
        effect.Reset();
        effect.PlaybackSpeed = stateInfo.speedMultiplier;
        effect.Play();
    }
}

File: Assets/Game/Lua/game/scene/actor/actor_trigger_effect.lua
ActorTriggerEffect = ActorTriggerEffect or BaseClass(ActorTriggerBase)

function ActorTriggerEffect:__init(anima_name, layer)
	self.anima_name = anima_name
	self.layer = layer
	self.transform = nil
	self.enabled = true
	self.target = nil

	self.effect_data = nil
end

function ActorTriggerEffect:__delete()
	self.effect_data = nil
end

-- 初始化预制体保存的配置数据(单个)
function ActorTriggerEffect:Init(effect_data)
	self.effect_data = effect_data
	self.delay = effect_data.triggerDelay
end

-- get/set
function ActorTriggerEffect:Enalbed(value)
	if value == nil then
		return self.enabled
	end
	self.enabled = value
end

function ActorTriggerEffect:OnEventTriggered(source, target, source_obj, target_obj, stateInfo)
	if self.enabled then
		self:OnEventTriggeredImpl(source, target, source_obj, target_obj, stateInfo)
	end
end

function ActorTriggerEffect:OnEventTriggeredImpl(source, target, source_obj, target_obj, stateInfo)
	local effect_data = self.effect_data
	if not effect_data then
		return
	end
	if effect_data.effectAsset and nil == next(effect_data.effectAsset) then
		return
	end
	-- Find the reference node.
	local reference = nil
	local deliverer = nil

	local reference_node = nil
	if effect_data.referenceNodeHierarchyPath ~= nil and effect_data.referenceNodeHierarchyPath ~= "" then
		reference_node = target.transform:Find(effect_data.referenceNodeHierarchyPath)
	end
	reference_node = reference_node or source.transform

	if effect_data.playerAtTarget then
		reference = target.transform
		deliverer = reference_node
	else
		reference = reference_node
		deliverer = reference_node
	end
	if reference == nil or deliverer == nil then
		return
	end

	local bundle_name = effect_data.effectAsset.BundleName
	local asset_name = effect_data.effectAsset.AssetName
	
	local async_loader = AllocAsyncLoader(self, "actor_trigger_effect")
	async_loader:SetIsUseObjPool(true)
	async_loader:SetObjAliveTime(5) --防止永久存在
	async_loader:Load(bundle_name, asset_name, function(obj)
		if IsNil(obj) then
			return
		end

		if self.layer then
			obj:SetLayerRecursively(self.layer)
		end
		
		local effect = obj:GetComponent(typeof(EffectControl))
		if effect == nil then
			async_loader:Destroy()
			return
		end

		if not IsNil(reference) and not IsNil(deliverer) then
			if effect_data.isAttach then
				effect.transform:SetParent(reference)
				if effect_data.isRotation then
					local direction = reference.position - deliverer.position
					direction.y = 0
					local temp_rotation = Vector3.zero
					if direction ~= Vector3.zero then
						temp_rotation = Quaternion.LookRotation(direction)
					end
					effect.transform:SetPositionAndRotation(reference.position, temp_rotation)
				else
					effect.transform.localPosition = Vector3.zero
					effect.transform.localRotation = Quaternion.identity
				end
				effect.transform.localScale = reference.localScale
			else
				effect.transform:SetPositionAndRotation(reference.position, reference.rotation)
				if effect.IsNoScalable then
					effect.transform.localScale = source.transform.lossyScale
				else
					effect.transform.localScale = reference.lossyScale
				end
			end
		end

		effect:Reset()
		effect:WaitFinsh(function()
			async_loader:Destroy()
		end)

		effect.enabled = true
		effect:Play()
	end)
end